### PR TITLE
Example Notebooks for Converting GP Models to TorchScript

### PR DESCRIPTION
**Pull request by jacobrgardner**
These examples demonstrate how GP models can be traced to ScriptModules and therefore used in production. At test time, the traced models are often up to 4x faster as well.

This should mostly address #1022 since I believe ScriptModules can directly be saved to ONNX format.

Additionally, cc/ @rangerbottle regarding #970, since converting your GP model to TorchScript *should* allow you to run it in LibTorch.

### CODE REVIEW

**Comment by jacobrgardner** [comment]
Our example notebook CI is failing because the notebooks don't check if cuda is available (it isn't in CI).

I'll put up a fix for that.

**Comment by gpleiss** [comment]
Can you add `if torch.cuda.is_available()` instead? That's what I did for the other notebooks.

**Comment by jacobrgardner** [comment]
Yep that's the plan, just been in meetings straight since putting this up.

File: examples/08_Advanced_Usage/TorchScript_Exact_Models.ipynb [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    "    \n",
+    "    def forward(self, x):\n",
+    "        output_dist = self.gp(x)\n",
+    "        return output_dist.mean, output_dist.variance\n",
+    "\n",
+    "with torch.no_grad(), gpytorch.settings.fast_pred_var():\n",
+    "    model.eval()\n",
+    "    test_x = torch.linspace(0, 1, 51)\n",
+    "    pred = model(test_x)  # Do precomputation\n",
+    "    traced_model = torch.jit.trace(MeanVarModelWrapper(model), test_x, check_trace=False)"
```
Comment (JasonXu12) [inline examples/08_Advanced_Usage/TorchScript_Exact_Models.ipynb]: Have you guys tried saving the traced model by "traced_model.save('model_state.pt')“ by any chance? Saving will cause error on my side due to "RuntimeError: Could not export Python function call 'RBFCovariance'. Remove calls to Python functions before export." 
We have the need for saving the .pt model and use it for C++ online inferences. Thank you and look forward to your reply!

Comment (jacobrgardner) [reply examples/08_Advanced_Usage/TorchScript_Exact_Models.ipynb]: Sorry for the delay finishing up this PR -- things are just getting a little busy as we approach the ICML deadline.

I'll take a look. What we may have to do is somehow avoid using the custom Functions for RBF and Matern when tracing.

**Comment by jacobrgardner** [comment]
Ended up adding a `gpytorch.settings.trace_mode` that we'll need to activate when tracing GPyTorch models for now until https://github.com/pytorch/pytorch/issues/22329 is resolved.

@JasonXu12 calling `traced_model.save('traced_model.pt')` now works.

**Comment by jacobrgardner** [comment]
@gpleiss this is all set. The `trace_mode` setting is I believe unavoidable. We're not allowed to do essentially anything lazy because we can't trace autograd Functions. This adds some minor code complexity to `VariationalStrategy` (and 2 of our kernels) but I think is worth it to have this functionality.

**Comment by JasonXu12** [comment]
Thank you for the reply@jacobrgardner.  As an external developer, my code relies on gpytorch v1.0.1, which doesn't include the trace_mode class yet and will cause error if I try to use trace_mode in ExactGPModel. Any workarounds or plans for the next release? Thank you!

**Comment by jacobrgardner** [comment]
@JasonXu12 Sure, once this gets merged in we can either release 1.0.2 since this is a particularly nice feature to have, or otherwise you can just install from github.

**Pull request merged**